### PR TITLE
Issue #4033 Percent Encoded Bad Requests

### PR DESCRIPTION
**Pull request by gregw**
The PR #4034 for #4033 has made request handling too lenient for requests.

### CODE REVIEW

**Comment by joakime** [comment]
Looks pretty straight forward.
Going to wait for the CI build/test first.

File: jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceAliasTest.java [hunk 1]
```diff
@@ -18,93 +18,153 @@
...
+            catch (MalformedURLException e)
+            {
+                assertTrue(true);
+            }
         }
-        catch (MalformedURLException e)
+        catch (InvalidPathException e)
         {
-            assertTrue(true);
+            // this file system does not suffer this problem
```
Comment (janbartel) [inline jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceAliasTest.java]: What does that comment mean? Does it mean that this exception will never be thrown?

File: tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml [hunk 2]
```diff
@@ -219,11 +219,5 @@
       <artifactId>test-container-initializer</artifactId>
       <version>${project.version}</version>
     </dependency>
```
Comment (janbartel) [inline tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml]: Why has the older version of the util jar been removed?

File: jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 3]
```diff
@@ -470,70 +470,24 @@ public static String decodePath(String path, int offset, int length)
...
-
-                        // lenient percent decoding
-                        if (i >= end)
-                        {
-                            // [LENIENT] a percent sign at end of string.
-                            builder.append('%');
-                            i = end;
-                        }
-                        else if (end > (i + 1))
+                        if ((i + 2) < end)
```
[the remark this answers is not in the document]
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: This was reverting previous code... so it is as it was before.

File: jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 4]
```diff
@@ -470,70 +470,24 @@ public static String decodePath(String path, int offset, int length)
...
-                                {
-                                    builder.append((byte)TypeUtil.parseInt(path, i + 1, 2, 16));
-                                    i += 2;
-                                }
-                                else
-                                {
-                                    builder.append(path, i, 3);
-                                    i += 2;
-                                }
+                                // TODO this is wrong. This is a codepoint not a char
```
[the remark this answers is not in the document]
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Yes, that TODO has been there for a long time.   We are more likely to remove %u support than fix it.

File: jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 5]
```diff
@@ -605,7 +567,8 @@ private static String decodeISO88591Path(String path, int offset, int length)
                         }
                         else
                         {
-                            builder.append((byte)(0xff & (TypeUtil.convertHexDigit(u) * 16 + TypeUtil.convertHexDigit(path.charAt(i + 2)))));
+                            // TODO this is wrong. 16 bits doesn't go to 8 reliably
```
[the remark this answers is not in the document]
Comment (gregw) [reply jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: yes, %u decoding is not perfect.

File: jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilTest.java [hunk 6]
```diff
@@ -117,6 +106,70 @@ public void testDecodePath(String encodedPath, String expectedPath)
...
+        arguments.add(Arguments.of("abc%uAB"));     // incomplete "%u####"
+        arguments.add(Arguments.of("abc%uA"));      // incomplete "%u####"
+        arguments.add(Arguments.of("abc%u"));       // incomplete "%u####"
+
+        // Test for null character (real world ugly test case)
+        // TODO is this a bad decoding or a bad URI ?
+        // arguments.add(Arguments.of("/%00/"));
+
+        // Deprecated Microsoft Percent-U encoding
+        // TODO still supported for now ?
```
[the remark this answers is not in the document]
Comment (gregw) [reply jetty-util/src/test/java/org/eclipse/jetty/util/URIUtilTest.java]: yes - reverting to previous state with TODOs

File: jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceAliasTest.java (hunk 1 above)
Comment (gregw) [reply jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceAliasTest.java]: @joakime ?

File: tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml (hunk 2 above)
Comment (gregw) [reply tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml]: I think this was something @joakime merged into the branch... it is an unrelated change.  I'll remove

File: jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceAliasTest.java (hunk 1 above)
Comment (joakime) [reply jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceAliasTest.java]: The original code ...

``` java
        File file0 = new File(__dir, "test.txt\0");	
        if (!file0.exists())	
            return;  // this file system does not suffer this problem
```

The updated code ...

``` java
try
{
    Path file0 = baseDir.resolve("test.txt\0");
} 
catch (InvalidPathException e) 
{
    // this file system does not suffer this problem
}
```

Comment (joakime) [reply jetty-util/src/test/java/org/eclipse/jetty/util/resource/ResourceAliasTest.java]: The entire test is moot if the filesystem it's running on catches the issue with files ending in null.
That's a test from https://bugs.eclipse.org/437800

**Pull request merged**